File: 309.best-time-to-buy-and-sell-stock-with-cooldown.py
#
# @lc app=leetcode id=309 lang=python3
#
# [309] Best Time to Buy and Sell Stock with Cooldown
#

# @lc code=start
import numpy as np
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # assuming the num of days is more than 2
        n = len(prices)
        value_mat = -1000 * np.ones((n, 2), dtype = int)
        value_mat[n - 1, 1] = prices[n - 1]
        value_mat[n - 1, 0] = 0
        value_mat[n - 2, 1] = max(value_mat[n - 1, 1], prices[n - 2])
        value_mat[n - 2, 0] = max(value_mat[n - 1, 0], value_mat[n - 1, 1] - prices[n - 2])
        for d in range(n - 3, -1 , -1):
            value_mat[d, 1] = max(value_mat[d + 2, 0]+ prices[d], value_mat[d + 1, 1])
            value_mat[d, 0] = max(value_mat[d + 1, 0], value_mat[d + 1, 1] - prices[d])

        return value_mat[0, 0]
            


        
# @lc code=end

File: 309.best-time-to-buy-and-sell-stock-with-cooldown.cpp
/*
 * @lc app=leetcode id=309 lang=cpp
 *
 * [309] Best Time to Buy and Sell Stock with Cooldown
 */

// @lc code=start
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if (n <= 1){
            return 0;
        }
        if ( n == 2){
            return max(0, prices[1] - prices[0]);
        }
        int profit[2][n];
        profit[1][n - 1] = prices[n - 1];
        profit[0][n - 1] = 0;
        profit[1][n - 2] = max(prices[n - 2], profit[1][n - 1]);
        profit[0][n - 2] = max(0 , -prices[n - 2] + profit[1][n - 1]);
        for(int i = n - 3; i >=0; --i){
            profit[1][i] = max(prices[i] + profit[0][i + 2], profit[1][i + 1]);
            profit[0][i] = max(- prices[i] + profit[1][i + 1], profit[0][i + 1]);
        }
        return profit[0][0];
    }
};
// @lc code=end

